File: 127. Word Ladder.py
'''
T - O(s^2*wordlist) for graph where s is len of string and substring function also takes s time so its s^2. The same would hold true for bfs 
S - O(s^2*wordlist) as we create intermediate string in graph and graph would hold all combinations of s and wordlist
Apprach - first graph is created for every string like grouping anagrams and then bfs is performed to reach endword

'''

from collections import defaultdict,deque
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        ladderMap = defaultdict(list)
        if endWord not in wordList:
            return 0
        for word in wordList:
            for w in range(len(word)):
                key = word[:w]+'*'+word[w+1:]
                ladderMap[key].append(word)
        size = 0
        Q = deque()
        Q.append([beginWord,1])
        visited = set()
        visited.add(beginWord)
        while len(Q) != 0:
            currWord,size = Q.popleft()
            for k in range(len(beginWord)):
                key= currWord[:k]+'*'+currWord[k+1:]
                for items in ladderMap[key]:
                    if items == endWord:
                        return size + 1
                    if items not in visited:
                        Q.append([items,size+1])
                        visited.add(items)
        return 0
                    
        

File: 127. Word Ladder.java
//Time complexity:O(n*m) m indicates length of wordlist n indicates length of word
//Space complexity:O(m)
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> set=new HashSet(wordList);
        if(!set.contains(endWord)){
            return 0;
        }
        Queue<String> q=new LinkedList();
        int level=1;
        q.add(beginWord);
        while(!q.isEmpty()){
            int size=q.size();
            for(int i=0;i<size;i++){
                String curr=q.poll();
                char[] currword=curr.toCharArray();
                for(int j=0;j<currword.length;j++){
                    char temp=currword[j];
                    for(char c='a';c<='z';c++){
                        if(currword[j]==c){
                            continue;
                        }
                        currword[j]=c;
                        String newword=String.valueOf(currword);
                        if(set.contains(newword)){
                            if(newword.equals(endWord)){
                                return level+1;
                            }
                            else{
                                q.add(newword);
                                set.remove(newword);
                            }
                        }
                        currword[j]=temp;
                    }
                }
            }
            level++;
        }
        return 0;
    }
}